icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+   * @return boolean value indication of the current content of the task - true indicates that `task_slot` holds a 
+   * previous result that should be handled by the user before writing the new task.
+   */
+  bool get_free_task(Task*& task_slot);
+
+  /**
+   * @brief Get task that holds previous result to be handled by the user. This function blocks the code until a 
+   * completed task is found or all tasks are idle with no result.
+   * @param completed_task - pointer to the internal task holding a previous result. a nullptr is 
+   * returned in case that all internal tasks are idle with no result (And as such no completed task will be found).
+   * Again if using completed_task to assign additional tasks, the existing result must be handled before hand.
+   */
+  void get_completed_task(Task*& completed_task);
+private:
+  /**
+   * @class Worker
+   * @brief the equivalent of a thread and additional data members required for executing tasks in parallel to main.
+   */
+  class Worker {
+  public:
+    /**
+     * @brief Constructor of `Worker`.
+     * Inits default values for the class's members and launches the thread.
+     */
+    Worker();
+    /**
+     * @brief Destructor of `Worker`.
+     * Signals the thread to terminate and joins it with main. The destructor does not handle existing tasks' results 
+     * and assumes the user have already handled all the results via `TasksManager`'s api.
+     */
+    ~Worker();
+
+    /**
+     * @brief function to be ran by the thread.
+     * Routinely checks for valid inputs in all of the worker's tasks. It executes valid tasks, later marking back that 
+     * the tasks are complete. This loops until a kill signal is sent via the class's destructor.
+     */
+    void run();
+
+    /**
+     * @brief Get free slot to insert new task to be executed. This isn't a blocking function - it checks all worker's 
+     * tasks and if no free one is found a nullptr is returned.
+     * @param task_slot - pointer to the internal task of the worker to allow the user to edit in the new task.
+     * @return boolean value indication of the current content of the task - true indicates that `task_slot` holds a 
+     * previous result that should be handled by the user before writing the new task.
+     */
+    bool get_free_task(Task*& task_slot);
+
+    /**
+     * @brief Get task that holds previous result to be handled by the user. This isn't a blocking function - it checks 
+     * all worker's tasks and if no completed one is found a nullptr is returned.
+     * @param completed_task - pointer to the internal task holding a previous result. 
+     * Again if using completed_task to assign additional tasks, the existing result must be handled before hand.
+     */
+    void get_completed_task(Task*& completed_task);
+
+    /**
+     * @brief Check if all worker's tasks are idle (free with no result to be handled)
+     */
+    bool are_all_idle();
```

Comment:
is_idle() instead?

Reply:
"is" referring to idle? because are_all refers to the tasks of the worker

Reply:
referring to the worker. Just a rename